x64: Refactor emission of LoadExtName

alexcrichton opened the pull request:
Don't use raw bytes when emitting this instruction but instead use
symbolic assembler directives to avoid needing to hardcode bytes. This
additionally fixes an issue where with `colocated` symbols the offset of
the symbol was not taken into account (but for Wasmtime it's always been
0 so this otherwise hasn't come up so far)

alexcrichton committed:
x64: Expand symbol-related tests

alexcrichton committed:
x64: Refactor emission of `LoadExtName`

Don't use raw bytes when emitting this instruction but instead use
symbolic assembler directives to avoid needing to hardcode bytes. This
additionally fixes an issue where with `colocated` symbols the offset of
the symbol was not taken into account (but for Wasmtime it's always been
0 so this otherwise hasn't come up so far)

cfallin reviewed (approved):
Looks reasonable; thanks! Good catch handling the offsets.

alexcrichton committed:
Fix emit tests

alexcrichton committed:
Update test expectation

alexcrichton committed:
Fuse `KnownOffsetTable` and `CodeSink` in the new assembler

Enables a `CodeSink` to know about all offsets, not just those against
labels. Resolves an issue where the fuzzing code sink was otherwise
unaware of `DeferredTarget::None`.

alexcrichton commented:
Ok to resolve the test failure I had to push up a semi-substantial change as the final commit here. cc @abrown on the assembler change, and mind taking another look @cfallin?

cfallin commented on cranelift/assembler-x64/src/api.rs:
```diff
@@ -24,17 +24,13 @@ pub trait CodeSink {
-    /// Return the byte offset of the current location in the code buffer;
-    /// required for assembling RIP-relative memory accesses.
-    fn current_offset(&self) -> u32;
+    /// Inform the code buffer of a use of `target` is about to happen within
```
grammar fix -- "inform ... of a use ... is about to happen" -> "inform ... that a use ... is about to happen"

cfallin commented on cranelift/assembler-x64/src/api.rs:
```diff
@@ -24,17 +24,13 @@ pub trait CodeSink {
-    fn current_offset(&self) -> u32;
+    /// Inform the code buffer of a use of `target` is about to happen within
+    /// this buffer at the current offset. After this method is called bytes are
+    /// then placed at this offset.
```
s/the bytes are then placed/the bytes are then expected to be placed/

(it's part of the contract for what the caller must do, rather than a result of this method call, right?)

cfallin commented on cranelift/codegen/src/isa/x64/inst/external.rs:
```diff
@@ -429,37 +429,59 @@ pub mod offsets {
+        match offset {
+            offsets::KEY_INCOMING_ARG => self.incoming_arg_offset,
+            offsets::KEY_SLOT_OFFSET => self.slot_offset,
+            _ => unreachable!(),
```
Maybe a panic here with a panic message including `offset`? Technically it's unreachable but the unreachability invariant is "remote" (part of another crate) so I'd rather make this easier debug / more verbose if we do miss something later...

cfallin reviewed (approved):
New approach looks reasonable, but deferring to @abrown for final opinions on the `CodeSink` refactor...

cfallin commented:
(Was this auto-added to merge queue? I gave r+ but Andrew should still weigh in and there were a few comment nits)

alexcrichton committed:
Review comments
